fsm_internal.h

```diff
@@ -0,0 +1,36 @@
+#pragma once
```

Comment:
These macros are kind of gross. I understand why they are needed but I kind of wish we have template functions or something. These should all be documented more thoroughly so first-years can follow.